text/0009-multilink_command_handling.md

```diff
@@ -0,0 +1,45 @@
+
+# Detailed Design
+
+The planned system requires changes in the QGC, Mavlink, and Autopilot Software. First in the Mavlink codebase a timestamp field is added to the following messages:
+
+  * COMMAND_INT
+  * COMMAND_LONG
+
+The timestamp field contains the vehicle time since the bootup. 
+
+In a seconds step the groundstation needs to be synchronized with the vehicle time. This can be achieved using the SYSTEM_TIME message. The groundstation tracks the system time for each vehicle and fills it in the timestamp field for every command which is sent. As long as no SYSTEM_TIME message is received the timestamp field is set to 0.
+
+Finally on the autopilot software a list, called "command list" from now on, is built to track all received command. Each item in that list contains the command id, based on the MAV_CMD enum, an the value of the timestamp field of the most recently received command with that id. On bootup the list is empty. It is filled once commands are received. For any received command it is evaluated if that command is accepted (passed to the system) or rejected based on the command list and the values of the command id and timestamp field according to the following decision tree:
+
+  * The timestamp field is 0:
+      * The command is always accepted.This ensures that commands from a groundstation, which is not synchronised yet, are accepted. In addition, by filling the timestamp field of the commands always with 0 all commands are accepted and the current behaviour can be kept.
+
+  * The timestamp field is nonzero:
+      * An item in the command list with the same command id as the received command exists:
+          * The value of the timestamp field of the received command is compared to the timestamp of the list item. If it is larger the command is accepted and value of the timestamp in the list item is updated. If it is smaller the command is rejected.
+
+      * No item in the command list with the same command id as the received command exists:
+          * The command is accepted and a new item in the command list with the values of the received command is generated.
+
+To save RAM the following mechanics can be implemented to limit the size of the command list:
+
+  * Remove items if the respective timestamp lies N seconds in the past, where N should be equal to the maximum expected latency of any connected link.
+  * Limit the list to M entries. If a command with a new id is received the item with the oldest timestamp is removed and an item with the details of the received command is inserted.
+
+# Alternatives
+
+  * Instead of the timestamp a counter could be used. However, it is unclear how multiple ground station can be supported
```

Comment:
This is done by associating it with the same link id tuple used for signing.

Reply:
Sure, when using the timestamp multiple ground stations can be supported. This comment was with regards to using a counter instead of a timestamp field. There I am not sure how multiple ground stations can be supported as you cannot determine if command number N from GCS 1 was sent before command number M of GCS 2.

In general I try to avoid using data from the signing as I would like to support Mavlink 1 and 2